src/raftstore/store/store.rs

```diff
@@ -1538,6 +1957,65 @@ impl<T: Transport, C: PdClient> Store<T, C> {
         }
     }
 
+    /// Check if a request is valid if it has valid prepare_merge/commit_merge proposal.
+    fn check_merge_proposal(&self, msg: &mut RaftCmdRequest) -> Result<()> {
+        if !msg.get_admin_request().has_prepare_merge()
+            && !msg.get_admin_request().has_commit_merge()
+        {
+            return Ok(());
+        }
+
+        let region_id = msg.get_header().get_region_id();
+        let peer = &self.region_peers[&region_id];
+        let region = peer.region();
+
+        if msg.get_admin_request().has_prepare_merge() {
+            let target_region = msg.get_admin_request().get_prepare_merge().get_target();
+            let peer = match self.region_peers.get(&target_region.get_id()) {
+                None => return Err(box_err!("target region doesn't exist.")),
+                Some(p) => p,
+            };
+            if peer.region() != target_region {
+                return Err(box_err!("target region not matched, skip proposing."));
+            }
+            if !util::is_sibling_regions(target_region, region) {
```

Comment:
What if these cases:
1) source region has a learner but the target not;
2) both have learners but not on the same store.

We should cancel the merge or just let learner destroy itself when it receive `merge/per-perge` proposals?

Reply:
Only when source/destination have the same voters and learners can they be merged.